File: src/services/api/apiKeys.ts
/**
 * API Keys management service
 * Uses TypeScript with @tauri-apps/plugin-fs for file operations
 * Only delegates to Rust for external API testing (IGDB, SteamGridDB)
 */
import { readTextFile, writeTextFile, exists } from "@tauri-apps/plugin-fs";
import { appConfigDir, join } from "@tauri-apps/api/path";
import { invoke } from "@tauri-apps/api/core";

const CONFIG_FILENAME = "api_keys.json";

export interface ApiKeysConfig {
  steamgriddbApiKey: string | null;
  igdbClientId: string | null;
  igdbClientSecret: string | null;
  steamApiKey: string | null;
  steamId: string | null;
  setupCompleted: boolean;
  hasSteamgriddb: boolean;
  hasIgdb: boolean;
  hasSteam: boolean;
}

export interface ApiKeysUpdateRequest {
  steamgriddbApiKey?: string | null;
  igdbClientId?: string | null;
  igdbClientSecret?: string | null;
  steamApiKey?: string | null;
  steamId?: string | null;
  markSetupCompleted?: boolean;
}

export interface ApiKeyTestResult {
  steamgriddbValid: boolean;
  steamgriddbMessage: string | null;
  igdbValid: boolean;
  igdbMessage: string | null;
  steamValid: boolean;
  steamMessage: string | null;
}

// Internal interface for file storage
interface StoredApiKeysConfig {
  steamgriddb_api_key?: string | null;
  igdb_client_id?: string | null;
  igdb_client_secret?: string | null;
  steam_api_key?: string | null;
  steam_id?: string | null;
  setup_completed?: boolean;
}

/**
 * Get the config file path
 */
async function getConfigPath(): Promise<string> {
  const configDir = await appConfigDir();
  return await join(configDir, CONFIG_FILENAME);
}

/**
 * Default configuration
 */
function getDefaultConfig(): ApiKeysConfig {
  return {
    steamgriddbApiKey: null,
    igdbClientId: null,
    igdbClientSecret: null,
    steamApiKey: null,
    steamId: null,
    setupCompleted: false,
    hasSteamgriddb: false,
    hasIgdb: false,
    hasSteam: false,
  };
}

/**
 * Check if a config has the specified API key
 */
function computeHasFlags(config: StoredApiKeysConfig): {
  hasSteamgriddb: boolean;
  hasIgdb: boolean;
  hasSteam: boolean;
} {
  const hasSteamgriddb = !!config.steamgriddb_api_key && config.steamgriddb_api_key.length > 0;
  const hasIgdb =
    !!config.igdb_client_id &&
    config.igdb_client_id.length > 0 &&
    !!config.igdb_client_secret &&
    config.igdb_client_secret.length > 0;
  const hasSteam =
    !!config.steam_api_key &&
    config.steam_api_key.length > 0 &&
    !!config.steam_id &&
    config.steam_id.length > 0;

  return { hasSteamgriddb, hasIgdb, hasSteam };
}

/**
 * Convert stored config (snake_case) to API config (camelCase)
 */
function convertToApiConfig(stored: StoredApiKeysConfig): ApiKeysConfig {
  const flags = computeHasFlags(stored);
  return {
    steamgriddbApiKey: stored.steamgriddb_api_key || null,
    igdbClientId: stored.igdb_client_id || null,
    igdbClientSecret: stored.igdb_client_secret || null,
    steamApiKey: stored.steam_api_key || null,
    steamId: stored.steam_id || null,
    setupCompleted: stored.setup_completed || false,
    hasSteamgriddb: flags.hasSteamgriddb,
    hasIgdb: flags.hasIgdb,
    hasSteam: flags.hasSteam,
  };
}

/**
 * Convert API config (camelCase) to stored config (snake_case)
 */
function convertToStoredConfig(config: ApiKeysConfig): StoredApiKeysConfig {
  return {
    steamgriddb_api_key: config.steamgriddbApiKey,
    igdb_client_id: config.igdbClientId,
    igdb_client_secret: config.igdbClientSecret,
    steam_api_key: config.steamApiKey,
    steam_id: config.steamId,
    setup_completed: config.setupCompleted,
  };
}

/**
 * Load API keys config from file
 */
export async function getApiKeys(): Promise<ApiKeysConfig> {
  try {
    const configPath = await getConfigPath();
    const fileExists = await exists(configPath);

    if (!fileExists) {
      console.info("API keys config file not found, using defaults");
      return getDefaultConfig();
    }

    const content = await readTextFile(configPath);
    const stored: StoredApiKeysConfig = JSON.parse(content);
    console.info("Loaded API keys config successfully");

    return convertToApiConfig(stored);
  } catch (error) {
    console.error("Failed to load API keys:", error);
    return getDefaultConfig();
  }
}

/**
 * Check if setup has been completed
 */
export async function needsSetup(): Promise<boolean> {
  try {
    const config = await getApiKeys();
    return !config.setupCompleted;
  } catch (error) {
    console.error("Failed to check setup status:", error);
    return true; // Default to true (setup needed) on error
  }
}

/**
 * Save API keys config to file
 */
export async function saveApiKeys(request: ApiKeysUpdateRequest): Promise<ApiKeysConfig> {
  try {
    const configPath = await getConfigPath();
    const currentConfig = await getApiKeys();

    // Apply updates
    const updated: ApiKeysConfig = {
      ...currentConfig,
      ...(request.steamgriddbApiKey !== undefined && {
        steamgriddbApiKey: request.steamgriddbApiKey,
      }),
      ...(request.igdbClientId !== undefined && {
        igdbClientId: request.igdbClientId,
      }),
      ...(request.igdbClientSecret !== undefined && {
        igdbClientSecret: request.igdbClientSecret,
      }),
      ...(request.steamApiKey !== undefined && {
        steamApiKey: request.steamApiKey,
      }),
      ...(request.steamId !== undefined && { steamId: request.steamId }),
      ...(request.markSetupCompleted && { setupCompleted: true }),
    };

    // Write to file
    const stored = convertToStoredConfig(updated);
    const content = JSON.stringify(stored, null, 2);
    await writeTextFile(configPath, content);

    console.info("Saved API keys config successfully");
    return updated;
  } catch (error) {
    console.error("Failed to save API keys:", error);
    throw error;
  }
}

/**
 * Skip initial setup (mark as completed without providing keys)
 */
export async function skipSetup(): Promise<void> {
  try {
    await saveApiKeys({ markSetupCompleted: true });
    console.info("Setup marked as completed (skipped)");
  } catch (error) {
    console.error("Failed to skip setup:", error);
    throw error;
  }
}

/**
 * Test API keys connectivity
 * This delegates to Rust because it requires complex HTTP operations
 */
export async function testApiKeys(request: ApiKeysUpdateRequest): Promise<ApiKeyTestResult> {
  try {
    return await invoke<ApiKeyTestResult>("test_api_keys", { request });
  } catch (error) {
    console.error("Failed to test API keys:", error);
    throw error;
  }
}

File: src-tauri/src/commands/api_keys.rs
use crate::services::{ApiKeysConfig, ApiKeysManager};
use serde::{Deserialize, Serialize};

/// API Keys configuration response (for frontend)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApiKeysResponse {
    pub steamgriddb_api_key: Option<String>,
    pub igdb_client_id: Option<String>,
    pub igdb_client_secret: Option<String>,
    pub steam_api_key: Option<String>,
    pub steam_id: Option<String>,
    pub setup_completed: bool,
    pub has_steamgriddb: bool,
    pub has_igdb: bool,
    pub has_steam: bool,
}

impl From<ApiKeysConfig> for ApiKeysResponse {
    fn from(config: ApiKeysConfig) -> Self {
        Self {
            has_steamgriddb: config.has_steamgriddb(),
            has_igdb: config.has_igdb(),
            has_steam: config.has_steam(),
            steamgriddb_api_key: config.steamgriddb_api_key,
            igdb_client_id: config.igdb_client_id,
            igdb_client_secret: config.igdb_client_secret,
            steam_api_key: config.steam_api_key,
            steam_id: config.steam_id,
            setup_completed: config.setup_completed,
        }
    }
}

/// API Keys update request (from frontend)
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApiKeysUpdateRequest {
    pub steamgriddb_api_key: Option<String>,
    pub igdb_client_id: Option<String>,
    pub igdb_client_secret: Option<String>,
    pub steam_api_key: Option<String>,
    pub steam_id: Option<String>,
    #[serde(default)]
    pub mark_setup_completed: bool,
}

#[derive(Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ApiKeyTestResult {
    pub steamgriddb_valid: bool,
    pub steamgriddb_message: Option<String>,
    pub igdb_valid: bool,
    pub igdb_message: Option<String>,
    pub steam_valid: bool,
    pub steam_message: Option<String>,
}

/// Get current API keys configuration
#[tauri::command]
pub async fn get_api_keys() -> Result<ApiKeysResponse, String> {
    log::info!("Getting API keys configuration...");

    let manager = ApiKeysManager::new().map_err(|e| e.to_string())?;
    let config = manager.load().map_err(|e| e.to_string())?;

    Ok(config.into())
}

/// Check if initial setup is required
#[tauri::command]
pub async fn needs_setup() -> Result<bool, String> {
    log::info!("Checking if setup is needed...");

    let manager = ApiKeysManager::new().map_err(|e| e.to_string())?;
    let needs_setup = !manager.is_setup_completed();

    log::info!("Setup needed: {}", needs_setup);
    Ok(needs_setup)
}

/// Save API keys configuration
#[tauri::command]
pub async fn save_api_keys(request: ApiKeysUpdateRequest) -> Result<ApiKeysResponse, String> {
    log::info!("Saving API keys configuration...");

    let manager = ApiKeysManager::new().map_err(|e| e.to_string())?;
    let mut config = manager.load().unwrap_or_default();

    if let Some(key) = request.steamgriddb_api_key {
        config.steamgriddb_api_key = if key.is_empty() { None } else { Some(key) };
    }
    if let Some(id) = request.igdb_client_id {
        config.igdb_client_id = if id.is_empty() { None } else { Some(id) };
    }
    if let Some(secret) = request.igdb_client_secret {
        config.igdb_client_secret = if secret.is_empty() { None } else { Some(secret) };
    }
    if let Some(key) = request.steam_api_key {
        config.steam_api_key = if key.is_empty() { None } else { Some(key) };
    }
    if let Some(id) = request.steam_id {
        config.steam_id = if id.is_empty() { None } else { Some(id) };
    }

    if request.mark_setup_completed {
        config.setup_completed = true;
    }

    manager.save(&config).map_err(|e| e.to_string())?;

    log::info!("API keys configuration saved successfully");
    log::info!(
        "  - SteamGridDB: {}, IGDB: {}, Steam: {}",
        if config.has_steamgriddb() { "✓" } else { "✗" },
        if config.has_igdb() { "✓" } else { "✗" },
        if config.has_steam() { "✓" } else { "✗" }
    );

    Ok(config.into())
}

/// Skip initial setup (mark as completed without providing keys)
#[tauri::command]
pub async fn skip_setup() -> Result<(), String> {
    log::info!("Skipping initial setup...");

    let manager = ApiKeysManager::new().map_err(|e| e.to_string())?;
    manager.mark_setup_completed().map_err(|e| e.to_string())?;

    log::info!("Setup marked as completed (skipped)");
    Ok(())
}

/// Test API keys connectivity
#[tauri::command]
pub async fn test_api_keys(request: ApiKeysUpdateRequest) -> Result<ApiKeyTestResult, String> {
    use crate::services::igdb::IGDBService;
    use crate::services::steamgriddb::SteamGridDBService;

    log::info!("Testing API keys...");

    let mut result = ApiKeyTestResult::default();

    // Test SteamGridDB
    if let Some(ref key) = request.steamgriddb_api_key {
        if !key.is_empty() {
            let service = SteamGridDBService::new(key.clone());
            match service.search_game("Portal").await {
                Ok(Some(_)) => {
                    result.steamgriddb_valid = true;
                    result.steamgriddb_message = Some("Connected successfully".to_string());
                }
                Ok(None) => {
                    result.steamgriddb_valid = true;
                    result.steamgriddb_message = Some("Connected (no test results)".to_string());
                }
                Err(e) => {
                    result.steamgriddb_message = Some(format!("Error: {}", e));
                }
            }
        }
    }

    // Test IGDB
    if let (Some(ref client_id), Some(ref client_secret)) = (
        request.igdb_client_id.clone(),
        request.igdb_client_secret.clone(),
    ) {
        if !client_id.is_empty() && !client_secret.is_empty() {
            let mut service = IGDBService::new(client_id.clone(), client_secret.clone());
            match service.authenticate().await {
                Ok(()) => {
                    result.igdb_valid = true;
                    result.igdb_message = Some("Authenticated successfully".to_string());
                }
                Err(e) => {
                    result.igdb_message = Some(format!("Auth error: {}", e));
                }
            }
        }
    }

    // Test Steam (just check format)
    if let (Some(ref key), Some(ref id)) = (request.steam_api_key.clone(), request.steam_id.clone())
    {
        if !key.is_empty() && !id.is_empty() {
            if key.len() >= 20 && id.chars().all(|c| c.is_ascii_digit()) {
                result.steam_valid = true;
                result.steam_message = Some("Format valid (not tested against API)".to_string());
            } else {
                result.steam_message = Some("Invalid format".to_string());
            }
        }
    }

    log::info!(
        "API key test results: SteamGridDB={}, IGDB={}, Steam={}",
        result.steamgriddb_valid,
        result.igdb_valid,
        result.steam_valid
    );

    Ok(result)
}
